sdk/eventhub/azure-eventhub/azure/eventhub/_tracing.py

```diff
@@ -0,0 +1,268 @@
+                        headers["traceparent"]
+                    )
+                    message = amqp_transport.update_message_app_properties(
+                        message,
+                        TRACE_PARENT_PROPERTY,
+                        headers["traceparent"]
+                    )
+                if "tracestate" in headers:
+                    message = amqp_transport.update_message_app_properties(
+                        message,
+                        TRACE_STATE_PROPERTY,
+                        headers["tracestate"]
+                    )
+
+                message_span.add_attribute(
+                    TraceAttributes.TRACE_NAMESPACE_ATTRIBUTE, TraceAttributes.TRACE_NAMESPACE)
+                message_span.add_attribute(
+                    TraceAttributes.TRACE_MESSAGING_SYSTEM_ATTRIBUTE, TraceAttributes.TRACE_MESSAGING_SYSTEM)
+
+                if additional_attributes:
+                    for key, value in additional_attributes.items():
+                        if value is not None:
+                            message_span.add_attribute(key, value)
+
+    except Exception as exp:  # pylint:disable=broad-except
+        _LOGGER.warning("trace_message had an exception %r", exp)
+
+    return message
+
+
+def get_span_links_from_received_events(events: Union[EventData, Iterable[EventData]]) -> List[Link]:
+    """Create span links from received events.
+
+    This will extract the traceparent and tracestate from the event properties and create span links
+    based on these values. The time the event was enqueued is also added as a link attribute.
+    """
+    # pylint:disable=isinstance-second-argument-not-valid-type
+    trace_events = events if isinstance(events, Iterable) else (events,)
+    links = []
+    try:
+        for event in trace_events:
+            headers = {}
+            if event.properties:
+                traceparent = (event.properties.get(TRACE_PARENT_PROPERTY, b"") or
+                               event.properties.get(TRACE_DIAGNOSTIC_ID_PROPERTY, b""))
+                if hasattr(traceparent, "decode"):
+                    traceparent = traceparent.decode(TRACE_PROPERTY_ENCODING)
+                if traceparent:
+                    headers["traceparent"] = traceparent
+
+                tracestate = event.properties.get(TRACE_STATE_PROPERTY, b"")
+                if hasattr(tracestate, "decode"):
+                    tracestate = tracestate.decode(TRACE_PROPERTY_ENCODING)
+                if tracestate:
+                    headers["tracestate"] = tracestate
+
+            enqueued_time = event.system_properties.get(PROP_TIMESTAMP)
+            attributes = {'enqueuedTime': enqueued_time} if enqueued_time else None
+            links.append(Link(headers, attributes=attributes))
+    except AttributeError:
+        pass
+    return links
+
+
+def get_span_links_from_batch(batch: EventDataBatch) -> List[Link]:
+    """Create span links from a batch of events."""
```

Comment:
@pvaneck - Just curious, but is this a new method you added/copied over from SB? B/c I can't find it anywhere in the current EH code 🤔

Reply:
This is used in the producer files. E.g. https://github.com/Azure/azure-sdk-for-python/pull/29934/files#diff-dccec39f540988e20599c82ec391628083564e8c45db43a64845b4c8b4214145R285